gtdynamics/factors/ContactPointFactor.h

```diff
@@ -0,0 +1,113 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file  ContactPointFactor.h
+ * @brief Factor to enforce point on link is in contact with environment point.
+ * @author: Varun Agrawal
+ */
+
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Vector.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+#include <string>
+
+#include "gtdynamics/utils/ContactPoint.h"
+
+namespace gtdynamics {
+
+/**
+ * ContactPointFactor is a two-way nonlinear factor which enforces a
+ * point on the link to be equal to the point of contact in the environment.
+ */
+class ContactPointFactor
+    : public gtsam::NoiseModelFactor2<gtsam::Pose3, gtsam::Point3> {
+ private:
+  using This = ContactPointFactor;
+  using Base = gtsam::NoiseModelFactor2<gtsam::Pose3, gtsam::Point3>;
+
+  gtsam::Point3 lPc_;  // The contact point in the link's COM frame.
+
+ public:
+  /**
+   * Constructor.
+   *
+   * @param link_pose_key Key for the CoM pose of the link in contact.
+   * @param point_key Key for the contact point in the environment.
+   * @param cost_model Noise model associated with this factor.
+   * @param lPc Static transform from point of contact to link CoM.
+   */
+  ContactPointFactor(gtsam::Key link_pose_key, gtsam::Key point_key,
+                     const gtsam::noiseModel::Base::shared_ptr &cost_model,
+                     const gtsam::Point3 &lPc)
```

Comment:
We have called this `point_com` consistently in, say, testPointGoalFactor.

Reply:
`lPc` is also a point in the spatial frame that the link comes in contact with. Made sense to follow the GTSAM terminology.

Reply:
I agree it makes sense, but it also makes the thing harder to understand. Looking at Daisha’s PR, we renamed it everywhere to “contact_in_com”. Could you please rename it for clarity?